Added file type accessor, and associated enums and helper methods.

Wardrop opened the pull request:
FreeImage automatically determines the file type, which is great, but ImageScience doesn't provide a means of accessing the determined file type. This adds that functionality.

Commit:
Allow reading file data from memory

Commit:
Optional argument for writing to memory buffer

Commit:
Simpler FIF lookup by passing full file name from ImageScience#save

Commit:
Cleanup extraneous test assertions and ImageScience#save method

Wardrop committed:
Added file type accessor, and associated enums and helper methods.

Wardrop committed:
Added gemspec

Wardrop committed:
Merge https://github.com/jcasts/image_science

Wardrop commented:
I also need the write to buffer functionality is one of the other closed pull requests that never got merged. I've manually merged that and it's been automatically included in this pull request.

Wardrop committed:
Added files to gemspec.

zenspider commented:
No gemspec please.

Wardrop committed:
Delete image_science.gemspec

Wardrop commented:
Done

zenspider commented on lib/image_science.rb:
```diff
@@ -58,6 +142,30 @@ class ImageScience
+  ##
+  # Returns an appropriate content type for the file type, e.g. 'image/jpeg'
+  def content_type(type = @file_type)
+    ImageScience::CONTENT_TYPES[ImageScience::FREE_IMAGE_FORMAT.key(type)]
```
Why doesn't this use `file_format`?

zenspider commented on lib/image_science.rb:
```diff
@@ -330,28 +449,48 @@ def cropped_thumbnail(size) # :yields: image
 
-    builder.c <<-"END"
-      VALUE save(char * output) {
+    builder.c_raw <<-"END"
```
Why `c_raw` ?

zenspider commented on lib/image_science.rb:
```diff
@@ -330,28 +449,48 @@ def cropped_thumbnail(size) # :yields: image
-          BOOL result = 0, unload = 0;
           GET_BITMAP(bitmap);
           flags = fif == FIF_JPEG ? JPEG_QUALITYSUPERB : 0;
+          BOOL result = 0, unload = 0;
```
I don't understand some of these diffs... is there a reason?

zenspider commented on lib/image_science.rb:
```diff
@@ -41,10 +124,11 @@ class ImageScience
+  # appropriate format.
   #
-  # :method: save
+  # :method: buffer
```
This is a confusing description now. It doesn't save (as far as the user is concerned)... just returns the (binary) string representation.

zenspider commented on test/test_image_science.rb:
```diff
@@ -9,20 +10,23 @@
 require 'minitest/unit'
 require 'minitest/autorun' if $0 == __FILE__
-require 'image_science'
+require_relative '../lib/image_science'
```
plain require please

zenspider commented on test/test_image_science.rb:
```diff
@@ -207,4 +211,30 @@ def test_rotate_not_45
+      end
+    end
+
+    file_data = File.new(@tmpjpeg).binmode.read
```
I'd prefer either `File.read` or block mode to ensure quick close.

zenspider commented on test/test_image_science.rb:
```diff
@@ -207,4 +211,30 @@ def test_rotate_not_45
+      end
+    end
+
+    file_data = File.new(@tmppath).binmode.read
```
Ditto

zenspider replied:
Oh. also, binmode probably doesn't work on 1.8... this won't be allowed until I officially drop 1.8

ETA: I can deal with that tho.

Wardrop committed:
Using binread in tests, changed require_relative to require, and now using file_format internally.

Wardrop commented:
I've addressed most of your concerns. Note, I've changed all instance of `.binmode.read` to `File::binread` which I know isn't supported in 1.8, but I've never targeted 1.8 so I'm unsure of what an appropriate alternative is? Some of your other issues raised were copied verbatim from pull #19, but I've fixed up some of those also. As for why pull #19 originally use `c_raw`, well changing it to `c` produces an error, but I'm not a C programmer so I wouldn't be able to tell you why; you have a much better chance of working that out :)

    ../gems/RubyInline-3.12.4/lib/inline.rb:491:in 'ruby2c': Unknown type "VALUE *" (ArgumentError)

zenspider commented on lib/image_science.rb:
```diff
@@ -13,6 +13,89 @@
+    FIF_RAW: 'image/raw',
+    FIF_WEBP: 'image/webp',
+    FIF_JXR: 'image/jxr'
+  }
```
I'm not against it... but this seems like a maintenance nightmare...

Wardrop replied:
Maybe you can come up with a way to infer this from FreeImage. I'm not a C developer so it would be out of my depth. I did find a couple of methods though in FreeImage that could probably be wrapped, such as `FreeImage_GetFIFMimeType` and `FreeImage_GetFIFFromMime`

zenspider commented:
No respnose. Closing.
